pkg/ring/kv/dynamodb/metrics.go

```diff
@@ -0,0 +1,93 @@
+package dynamodb
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	grpcUtils "github.com/weaveworks/common/grpc"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/instrument"
+)
+
+type dynamodbInstrumentation struct {
+	kv         dynamodbKV
+	ddbMetrics *dynamodbMetrics
+}
+
+type dynamodbMetrics struct {
+	dynamodbRequestDuration *instrument.HistogramCollector
+	dynamodbUsageMetrics    *prometheus.CounterVec
+}
+
+func newDynamoDbMetrics(registerer prometheus.Registerer) *dynamodbMetrics {
+	dynamodbRequestDurationCollector := instrument.NewHistogramCollector(promauto.With(registerer).NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "dynamodb_kv_request_duration_seconds",
+		Help:    "Time spent on dynamodb requests.",
+		Buckets: prometheus.DefBuckets,
+	}, []string{"operation", "status_code"}))
+
+	dynamodbUsageMetrics := promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+		Name: "dynamodb_kv_read_capacity",
+		Help: "Total used read capacity on dynamodb",
+	}, []string{"operation"})
+
+	dynamodbMetrics := dynamodbMetrics{
+		dynamodbRequestDuration: dynamodbRequestDurationCollector,
+		dynamodbUsageMetrics:    dynamodbUsageMetrics,
+	}
+	return &dynamodbMetrics
+}
+
+func (d dynamodbInstrumentation) List(ctx context.Context, key dynamodbKey) ([]string, float64, error) {
+	var resp []string
+	var totalCapacity float64
+	err := instrument.CollectedRequest(ctx, "List", d.ddbMetrics.dynamodbRequestDuration, errorCode, func(ctx context.Context) error {
+		var err error
+		resp, totalCapacity, err = d.kv.List(ctx, key)
+		return err
+	})
+	d.ddbMetrics.dynamodbUsageMetrics.WithLabelValues("List").Add(totalCapacity)
```

Comment:
I am not familiar with DDB. May I know what does the capacity mean here? Like the quota we just used for querying? @danielblando

Reply:
Correct, the capacity is the quota used for one execution to the API. It depends on the amount of data returned.
There is a better explanation here https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadWriteCapacityMode.html

With this metrics we can have a ideia of how much usage we are spending in ddb and which action is heavier

Reply:
Cool this makes sense. At the beginning I am not sure if this means the capacity I used for that read request or the available capacity I can still use.
After reading the doc you shared this makes to me now. Thanks!